### PR TITLE
net/dnscache: don't cancel the TLS context before writing to the result channel.

### PR DESCRIPTION
**Pull request by danderson**
Cancelling the context makes the timeout goroutine race with the write that
reports a successful TLS handshake, so you can end up with a successful TLS
handshake that mysteriously reports that it timed out after ~0s in flight.

The context is always canceled and cleaned up as the function exits, which
happens mere microseconds later, so just let function exit clean up and
thereby avoid races.

Signed-off-by: David Anderson <danderson@tailscale.com>

### CODE REVIEW

File: net/dnscache/dnscache.go [hunk 1]
```diff
@@ -457,9 +457,7 @@ func TLSDialer(fwd DialContextFunc, dnsCache *Resolver, tlsConfigBase *tls.Confi
 			}
 		}()
 		go func() {
-			err := tlsConn.Handshake()
-			handshakeTimeoutCancel()
-			errc <- err
+			errc <- tlsConn.Handshake()
```
Comment (bradfitz) [inline net/dnscache/dnscache.go:460]: really all this crap can be simplified now that this is in std as of Go 1.17: https://pkg.go.dev/crypto/tls#Conn.HandshakeContext

File: net/dnscache/dnscache.go [hunk 2]
```diff
@@ -457,9 +457,7 @@ func TLSDialer(fwd DialContextFunc, dnsCache *Resolver, tlsConfigBase *tls.Confi
 			}
 		}()
 		go func() {
-			err := tlsConn.Handshake()
-			handshakeTimeoutCancel()
```
Comment (bradfitz) [inline net/dnscache/dnscache.go:461]: ah! I see. it made the L452 goroutine readable and send to errc first sometimes.

**Review by bradfitz — APPROVED** [review]
LGTM. Up to you to switch to HandshakeContext in this PR or a future one.

**Comment by DentonGentry** [comment]
Would we want this in 1.20.1?

If the functionality works and just leads to annoying-to-us logging, that isn't as important to get into 1.20.1.

File: net/dnscache/dnscache.go (hunk 2 above)
Comment (danderson) [reply net/dnscache/dnscache.go:461]: Yup, exactly.

File: net/dnscache/dnscache.go (hunk 1 above)
Comment (danderson) [reply net/dnscache/dnscache.go]: Replaced with HandshakeContext.

**Comment by danderson** [comment]
It's probably not necessary for 1.20.1. In production code, the only effect is that we might have to retry ~once to get a working control or DERP connection, because we mistakenly judge a working connection to be bad. It makes tests flaky, but doesn't hugely affect the operation of tailscaled, since we retry everything with backoff anyway.

**Review by bradfitz — APPROVED** [review]

**Comment by bradfitz** [comment]
> Would we want this in 1.20.1?
>
> If the functionality works and just leads to annoying-to-us logging, that isn't as important to get into 1.20.1.

I'd be fine (and slightly supportive) of Dave's first rev of this PR getting backported, but not the bigger change to use HandshakeContext. That I'd rather let bake a bit.

**Comment by danderson** [comment]
Well, I just wiped that first rev to replace it with HandshakeContext :). I don't think anyone would notice this bug outside of tests, so I'm meh on backporting, but sounds like on balance everyone's neutral or in favor. I still have the old commit in my reflog somewhere, I'll arrange this as 2 commits. Stand by.

**Comment by bradfitz** [comment]
LGTM as 2 commits, thanks!

**Comment by danderson** [comment]
I've cherry-picked commit a9da6b73 to `release-branch/1.20`, so it'll be in the next patch release.

**Pull request merged**